tests/basic/replay.rs

```diff
@@ -178,3 +178,115 @@ fn replay_persist_none() {
     // All our current failure persistence modes print the word "schedule", so check that's missing
     assert!(!output.contains("schedule"));
 }
+
+/// Tests that events not causally related to the failure are never scheduled.
+#[test]
+fn replay_causality() {
+    // The main thread will spawn three threads:
+    // - A, which acquires the lock and sets it to one;
+    // - B, which acquires the lock and asserts it is zero;
+    // - C, which sets an unrelated atomic Boolean.
+    // If A runs before B (as in the schedule below), then the assertion
+    // fails. If we provide the clock of this failure to the scheduler, we
+    // should never see thread C do anything, i.e., the atomic Boolean should
+    // never be set, because it is not causally related to the actual panic.
+
+    use std::sync::atomic::{AtomicBool, Ordering};
+
+    let flag = Arc::new(AtomicBool::new(false));
+    let flag_clone = Arc::clone(&flag);
+
+    let result = panic::catch_unwind(|| {
+        let schedule = Schedule::new_from_task_ids(0, vec![0, 1, 0, 2, 0, 3, 0, 1, 1, 2, 2]);
+        let mut scheduler = ReplayScheduler::new_from_schedule(schedule);
+        scheduler.set_target_clock(&[2, 2, 1]);
+        let mut config = Config::new();
+        config.failure_persistence = FailurePersistence::None;
+        let runner = Runner::new(scheduler, config);
+        runner.run(move || {
+            let lock = Arc::new(Mutex::new(0usize));
+            let lock_clone = Arc::clone(&lock);
+            thread::spawn(move || {
+                *lock_clone.lock().unwrap() = 1;
+            });
+            thread::spawn(move || {
+                let guard = lock.lock().unwrap();
+                assert!(*guard == 0, "expected panic");
+                drop(guard);
+            });
+            let flag_clone = Arc::clone(&flag_clone);
+            thread::spawn(move || {
+                // Note that this operation is performed in a separate thread,
+                // since the (non-Shuttle) atomic does not increment the clock
+                // of the current thread. If the atomic were set instead in the
+                // main thread, then the clocks of "setting the atomic" and "B
+                // acquiring a lock" would be indistinguishable. However, we
+                // need the non-Shuttle atomic to smuggle data out of a this
+                // panicking test.
+                flag_clone.store(true, Ordering::SeqCst);
+            });
+        });
+    })
+    .expect_err("test should panic");
+    let output = result.downcast::<&str>().unwrap();
+    assert_eq!(*output, "expected panic");
```

Comment:
Could you also add an assert about `steps_skipped` to test that functionality too?

Reply:
I'm not sure how to do this (easily): `Runner::new` takes the scheduler we construct, so there is no access to it afterwards.

Reply:
Okay for now. Maybe we can find a way to get additional metrics through the metrics scheduler